File: axiom_engine/claude_backend.py
"""Claude (Anthropic Messages API) generation backend.

Optional adapter that replaces the local Axiom-TTT generation path with
Anthropic-hosted Claude. Useful when:

* The local TTT engine has no trained checkpoint and emits gibberish
  via the SHA-256-hashing ``SimpleTokenizer`` — Claude provides real
  text while the rest of the Axiom HTTP surface (``/v1/models``,
  ``/v1/chat/completions``, ``/v1/messages``) keeps the same protocol.
* You want to keep client integrations on Axiom's wire format while
  swapping the underlying generator.

Important caveat
----------------
When this backend is active, ``/v1/adapt`` and the W_tilde lifecycle
become **no-ops with respect to actual generation** — Claude is a
remote frozen model and we cannot push gradient updates into it.
Sessions still exist (so checkpoint round-trips and session IDs
remain compatible), but they no longer influence output.

Activation
----------
The backend is opt-in via ``AXIOM_BACKEND=claude`` plus the standard
``ANTHROPIC_API_KEY`` (or by constructing ``ClaudeBackend`` directly
and passing it to the server). When inactive, generation falls back
to the local InferencePipeline path.
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from typing import List, Optional


DEFAULT_CLAUDE_MODEL = "claude-haiku-4-5-20251001"


@dataclass
class ClaudeMessage:
    """Minimal chat message shape consumed by the backend."""

    role: str
    content: str


@dataclass
class ClaudeBackend:
    """Anthropic Messages API backend for the Axiom server.

    Parameters
    ----------
    model:
        Anthropic model identifier. Defaults to a current Haiku snapshot.
    api_key:
        Optional override; falls back to the ``ANTHROPIC_API_KEY`` env var.
    default_system:
        System prompt injected when the request does not supply one.
    """

    model: str = DEFAULT_CLAUDE_MODEL
    api_key: Optional[str] = None
    default_system: Optional[str] = None
    _client: object = field(default=None, init=False, repr=False)

    def __post_init__(self) -> None:
        try:
            from anthropic import Anthropic
        except ImportError as exc:
            raise RuntimeError(
                "ClaudeBackend requires the 'anthropic' package: "
                "pip install anthropic"
            ) from exc

        api_key = self.api_key or os.environ.get("ANTHROPIC_API_KEY")
        if not api_key:
            raise RuntimeError(
                "ClaudeBackend requires an API key — set ANTHROPIC_API_KEY "
                "or pass api_key=... explicitly."
            )
        self._client = Anthropic(api_key=api_key)

    # ------------------------------------------------------------------
    # Generation entry points used by the server
    # ------------------------------------------------------------------

    def generate(self, prompt: str, max_tokens: int) -> str:
        """Single-turn completion from a bare prompt string."""
        return self.generate_chat(
            [ClaudeMessage(role="user", content=prompt)],
            max_tokens=max_tokens,
            system=self.default_system,
        )

    def generate_chat(
        self,
        messages: List[ClaudeMessage],
        max_tokens: int,
        system: Optional[str] = None,
    ) -> str:
        """Multi-turn completion. Non-user/assistant roles are folded into system."""
        anth_messages, folded_system = _normalize_messages(messages)
        effective_system = system or folded_system or self.default_system

        kwargs = {
            "model": self.model,
            "max_tokens": max_tokens,
            "messages": anth_messages,
        }
        if effective_system:
            kwargs["system"] = effective_system

        response = self._client.messages.create(**kwargs)
        return _extract_text(response)


# ----------------------------------------------------------------------
# Helpers
# ----------------------------------------------------------------------


def _normalize_messages(
    messages: List[ClaudeMessage],
) -> tuple[list[dict], Optional[str]]:
    """Split out system messages and coerce alternating user/assistant turns.

    Anthropic requires the conversation in ``messages`` to begin with a
    ``user`` turn and only contain ``user``/``assistant`` roles. System
    prompts are passed via the top-level ``system`` field. We fold any
    ``system`` messages (or unknown roles) into a single system string.
    """
    system_parts: list[str] = []
    chat: list[dict] = []
    for msg in messages:
        if msg.role == "system":
            system_parts.append(msg.content)
        elif msg.role in ("user", "assistant"):
            chat.append({"role": msg.role, "content": msg.content})
        else:
            system_parts.append(f"[{msg.role}] {msg.content}")

    if not chat:
        chat = [{"role": "user", "content": ""}]
    elif chat[0]["role"] != "user":
        chat.insert(0, {"role": "user", "content": ""})

    folded = "\n\n".join(p for p in system_parts if p) or None
    return chat, folded


def _extract_text(response: object) -> str:
    """Pull plain text out of an Anthropic Messages response."""
    parts: list[str] = []
    for block in getattr(response, "content", []) or []:
        if getattr(block, "type", None) == "text":
            parts.append(getattr(block, "text", ""))
    return "".join(parts)


# ----------------------------------------------------------------------
# Environment-driven factory
# ----------------------------------------------------------------------


def backend_from_env() -> Optional[ClaudeBackend]:
    """Return a configured backend when ``AXIOM_BACKEND=claude``, else None."""
    if os.environ.get("AXIOM_BACKEND", "").lower() != "claude":
        return None
    model = os.environ.get("AXIOM_CLAUDE_MODEL", DEFAULT_CLAUDE_MODEL)
    system = os.environ.get("AXIOM_CLAUDE_SYSTEM") or None
    return ClaudeBackend(model=model, default_system=system)

File: axiom_engine_rs/src/claude_backend.rs
//! Anthropic Messages API client used as an optional generation backend.
//!
//! When configured (via [`ClaudeBackend::from_env`] or constructed directly
//! and installed on [`crate::server::AppState`]), the server routes
//! generation requests through Anthropic's hosted Claude API instead of
//! the local Axiom-TTT pipeline. The OpenAI and Anthropic HTTP surfaces
//! continue to respond in their native shapes — only the underlying
//! generator changes.
//!
//! Caveat: with this backend active, `/v1/adapt` and the per-layer
//! W̃ lifecycle become no-ops with respect to actual generation because
//! Claude is a remote frozen model. Sessions still exist so the wire
//! contract holds, but adaptation no longer influences output.

use std::time::Duration;

use reqwest::blocking::Client;
use serde::{Deserialize, Serialize};

pub const DEFAULT_CLAUDE_MODEL: &str = "claude-haiku-4-5-20251001";
const ANTHROPIC_VERSION: &str = "2023-06-01";
const DEFAULT_BASE_URL: &str = "https://api.anthropic.com";

#[derive(Debug, Clone)]
pub struct ChatTurn {
    pub role: String,
    pub content: String,
}

#[derive(Debug, Clone)]
pub struct ClaudeBackend {
    model: String,
    api_key: String,
    base_url: String,
    default_system: Option<String>,
    client: Client,
}

impl ClaudeBackend {
    pub fn new(model: String, api_key: String, default_system: Option<String>) -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(120))
            .build()
            .expect("reqwest blocking client should construct");
        Self {
            model,
            api_key,
            base_url: DEFAULT_BASE_URL.to_string(),
            default_system,
            client,
        }
    }

    /// Construct a backend from environment variables, returning `None`
    /// when ``AXIOM_BACKEND`` is not set to ``claude``.
    pub fn from_env() -> Option<Self> {
        if std::env::var("AXIOM_BACKEND")
            .map(|v| v.to_lowercase())
            .ok()
            .as_deref()
            != Some("claude")
        {
            return None;
        }
        let api_key = std::env::var("ANTHROPIC_API_KEY").ok()?;
        let model =
            std::env::var("AXIOM_CLAUDE_MODEL").unwrap_or_else(|_| DEFAULT_CLAUDE_MODEL.to_string());
        let default_system = std::env::var("AXIOM_CLAUDE_SYSTEM").ok();
        let mut backend = Self::new(model, api_key, default_system);
        if let Ok(base) = std::env::var("ANTHROPIC_BASE_URL") {
            backend.base_url = base;
        }
        Some(backend)
    }

    pub fn model(&self) -> &str {
        &self.model
    }

    /// Bare prompt → response text.
    pub fn generate(&self, prompt: &str, max_tokens: usize) -> Result<String, String> {
        self.generate_chat(
            &[ChatTurn {
                role: "user".to_string(),
                content: prompt.to_string(),
            }],
            max_tokens,
            None,
        )
    }

    /// Multi-turn chat → response text. Non-user/assistant roles are
    /// folded into the system field.
    pub fn generate_chat(
        &self,
        turns: &[ChatTurn],
        max_tokens: usize,
        system: Option<String>,
    ) -> Result<String, String> {
        let (messages, folded_system) = normalize_turns(turns);
        let effective_system = system
            .or(folded_system)
            .or_else(|| self.default_system.clone());

        let body = AnthropicRequest {
            model: &self.model,
            max_tokens,
            messages: &messages,
            system: effective_system.as_deref(),
        };

        let url = format!("{}/v1/messages", self.base_url.trim_end_matches('/'));
        let response = self
            .client
            .post(&url)
            .header("x-api-key", &self.api_key)
            .header("anthropic-version", ANTHROPIC_VERSION)
            .header("content-type", "application/json")
            .json(&body)
            .send()
            .map_err(|e| format!("anthropic request failed: {e}"))?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response
                .text()
                .unwrap_or_else(|_| "<no body>".to_string());
            return Err(format!("anthropic API error {status}: {text}"));
        }

        let parsed: AnthropicResponse = response
            .json()
            .map_err(|e| format!("anthropic response decode failed: {e}"))?;

        Ok(parsed
            .content
            .into_iter()
            .filter(|block| block.block_type == "text")
            .map(|block| block.text)
            .collect::<Vec<_>>()
            .concat())
    }
}

#[derive(Serialize)]
struct AnthropicRequest<'a> {
    model: &'a str,
    max_tokens: usize,
    messages: &'a [AnthropicChatMessage],
    #[serde(skip_serializing_if = "Option::is_none")]
    system: Option<&'a str>,
}

#[derive(Serialize)]
struct AnthropicChatMessage {
    role: String,
    content: String,
}

#[derive(Deserialize)]
struct AnthropicResponse {
    #[serde(default)]
    content: Vec<AnthropicContentBlock>,
}

#[derive(Deserialize)]
struct AnthropicContentBlock {
    #[serde(rename = "type")]
    block_type: String,
    #[serde(default)]
    text: String,
}

/// Split system turns out and coerce the chat sequence to start with `user`
/// (Anthropic requires alternating user/assistant beginning with user).
fn normalize_turns(turns: &[ChatTurn]) -> (Vec<AnthropicChatMessage>, Option<String>) {
    let mut system_parts: Vec<String> = Vec::new();
    let mut chat: Vec<AnthropicChatMessage> = Vec::new();
    for turn in turns {
        match turn.role.as_str() {
            "system" => system_parts.push(turn.content.clone()),
            "user" | "assistant" => chat.push(AnthropicChatMessage {
                role: turn.role.clone(),
                content: turn.content.clone(),
            }),
            other => system_parts.push(format!("[{other}] {}", turn.content)),
        }
    }

    if chat.is_empty() {
        chat.push(AnthropicChatMessage {
            role: "user".to_string(),
            content: String::new(),
        });
    } else if chat[0].role != "user" {
        chat.insert(
            0,
            AnthropicChatMessage {
                role: "user".to_string(),
                content: String::new(),
            },
        );
    }

    let folded = if system_parts.is_empty() {
        None
    } else {
        Some(system_parts.join("\n\n"))
    };
    (chat, folded)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normalize_extracts_system_messages() {
        let turns = vec![
            ChatTurn {
                role: "system".into(),
                content: "be terse".into(),
            },
            ChatTurn {
                role: "user".into(),
                content: "hi".into(),
            },
            ChatTurn {
                role: "assistant".into(),
                content: "hello".into(),
            },
            ChatTurn {
                role: "user".into(),
                content: "again".into(),
            },
        ];
        let (chat, system) = normalize_turns(&turns);
        assert_eq!(system.as_deref(), Some("be terse"));
        assert_eq!(chat.len(), 3);
        assert_eq!(chat[0].role, "user");
        assert_eq!(chat[1].role, "assistant");
        assert_eq!(chat[2].role, "user");
    }

    #[test]
    fn normalize_forces_user_first_turn() {
        let turns = vec![ChatTurn {
            role: "assistant".into(),
            content: "ok".into(),
        }];
        let (chat, _) = normalize_turns(&turns);
        assert_eq!(chat[0].role, "user");
        assert_eq!(chat[1].role, "assistant");
    }

    #[test]
    fn from_env_disabled_when_backend_unset() {
        std::env::remove_var("AXIOM_BACKEND");
        assert!(ClaudeBackend::from_env().is_none());
    }
}
